cmd/gwi-api/api/api.go

```diff
@@ -0,0 +1,321 @@
+		return echo.NewHTTPError(http.StatusBadRequest, "email is required")
+	case volunteer.Password == "":
+		return echo.NewHTTPError(http.StatusBadRequest, "password is required")
+	}
+
+	// Call the store.
+	ctx := c.Request().Context()
+	if err := api.store.AuthenticateOrCreateVolunteerAccount(ctx, &volunteer); err != nil {
+		if err == ErrPasswordMismatch {
+			return echo.ErrUnauthorized
+		}
+		return err
+	}
+
+	// Create a JWT token and return it.
+	token, err := api.GenerateJWTToken(&volunteer)
+	if err != nil {
+		return err
+	}
+
+	return c.JSONPretty(http.StatusOK, echo.Map{
+		"token": token,
+	}, "  ")
+}
+
+// GetMe returns account details for the currently authenticated volunteer.
+//
+// Returns:
+//   * OK on success
+//
+// NOTE: Not documenting other methods for time reasons.
+func (api *API) GetMe(c echo.Context) error {
+	// Get email for the current user.
+	email, ok := api.GetEmailFromJWTToken(c)
+	if !ok {
+		return echo.ErrUnauthorized
+	}
+
+	// Load volunteer from the store.
+	ctx := c.Request().Context()
+	volunteer, err := api.store.GetVolunteerByEmail(ctx, email)
+	if err != nil {
+		return err
+	}
+
+	// Return the resource.
+	return c.JSONPretty(http.StatusOK, volunteer, "  ")
+}
+
+func (api *API) PostTeam(c echo.Context) error {
+	// Decode the resource.
+	var team Team
+	if err := c.Bind(&team); err != nil {
+		return err
+	}
+
+	// Validate the body.
+	switch {
+	case team.ID == "":
+		return echo.NewHTTPError(http.StatusBadRequest, "id is required")
+	}
+
+	// Call the store.
+	ctx := c.Request().Context()
+	if err := api.store.CreateTeam(ctx, &team); err != nil {
+		if err == ErrAlreadyExists {
+			return echo.NewHTTPError(http.StatusConflict)
+		}
+		return err
+	}
+
+	// Returning the resource created is the usual practice.
+	return c.JSONPretty(http.StatusCreated, &team, "  ")
+}
+
+func (api *API) GetTeamByID(c echo.Context) error {
+	// Load team from the store.
+	teamID := c.Param("id")
+	ctx := c.Request().Context()
+	team, err := api.store.GetTeamByID(ctx, teamID)
+	if err != nil {
+		if err == ErrNotFound {
+			return echo.ErrNotFound
+		}
+		return err
+	}
+
+	// Return the resource.
+	return c.JSONPretty(http.StatusOK, team, "  ")
+}
+
+func (api *API) GetTeamMembers(c echo.Context) error {
+	// Consult the store.
+	teamID := c.Param("id")
+	ctx := c.Request().Context()
+	members, err := api.store.ListTeamMembers(ctx, teamID)
+	if err != nil {
+		return err
+	}
+
+	// Return team members. Make sure the slice is not nil to return [] on no member.
+	if members == nil {
+		members = []Volunteer{}
+	}
+	return c.JSONPretty(http.StatusOK, members, "  ")
+}
+
+func (api *API) PutTeamMemberByEmail(c echo.Context) error {
+	teamID := c.Param("id")
+	email := c.Param("email")
+	ctx := c.Request().Context()
+	if err := api.store.AddTeamMember(ctx, teamID, email); err != nil {
+		switch err {
+		case ErrNotFound:
+			return echo.ErrNotFound
+		case ErrAlreadyExists:
+			return nil
```

Comment:
Could you point me to see how this `nil` is handled?

Reply:
`200 OK` is written by `net/http.Server` when no other response is written.

https://golang.org/src/net/http/server.go#L1583

But I have to say that it might be more safe to do `return c.NoContent(http.StatusOK)` since it is better not to depend on the caller, in this case `net/http`. Then you decide to use another HTTP server that for some reason does something else, you are doomed. I mean, it's pretty much an edge case, but better safe than sorry...